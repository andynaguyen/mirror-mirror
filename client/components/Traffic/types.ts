import { QueryProps } from 'react-apollo';
import { TrafficData } from '../../../common/types';

interface TrafficProps extends ApolloWrapperProps {
  data?: QueryProps<TrafficData>;
  traffic: TrafficData;
}

interface ApolloWrapperProps {
  visible: boolean;
}

type TrafficResponse = {
  traffic: TrafficData;
};

export { TrafficProps, ApolloWrapperProps, TrafficResponse };
