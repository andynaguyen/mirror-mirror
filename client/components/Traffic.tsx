import axios from 'axios';
import * as React from 'react';
import { graphql, QueryProps } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import withRefresh from '../refresher';

const query = gql`
  query {
    traffic {
      distance
      duration
    }
  }
`;

const REFRESH_INTERVAL = 1000 * 60 * 15; // 15 min

interface TrafficProps extends ApolloWrapperProps {
  data: QueryProps<TrafficData>;
  traffic: TrafficData;
}

interface ApolloWrapperProps {
  visible: boolean;
}

type TrafficData = {
  distance: string;
  duration: string;
};

type Response = {
  traffic: TrafficData;
};

const Container = styled.div`
  display: flex;
  justify-content: flex-end;
  margin-top: 8px;
`;

const CarIcon = styled.i`
  color: white;
  font-size: 22px;
  width: 36px;
`;

const Traffic: React.SFC<TrafficProps> = ({ traffic, visible }) => {
  return visible ? (
    <Container>
      <CarIcon className="icon-directions_car" />
      {traffic.distance} - {traffic.duration}
    </Container>
  ) : null;
};

const defaultState = {
  distance: 'N/A',
  duration: 'N/A',
};

/**
 * A component showing basic traffic conditions updated at every 15 minutes.
 */
const EnhancedTraffic: React.ComponentClass<ApolloWrapperProps> = graphql<Response, ApolloWrapperProps>(query, {
  options: {
    pollInterval: REFRESH_INTERVAL,
    fetchPolicy: 'network-only',
  },
  props: ({ data: { traffic, error } }) =>
    error
      ? { traffic: defaultState }
      : {
          traffic: {
            distance: traffic && traffic.distance,
            duration: traffic && traffic.duration,
          },
        },
})(Traffic);
export default EnhancedTraffic;
