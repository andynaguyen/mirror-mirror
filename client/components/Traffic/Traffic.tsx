import * as React from 'react';
import styled from 'styled-components';
import { TrafficProps } from './types';

const Container = styled.div`
  display: flex;
  justify-content: flex-end;
  margin-top: 8px;
`;

const CarIcon = styled.i`
  color: white;
  font-size: 22px;
  width: 36px;
`;

export default ({ traffic, visible }: TrafficProps) => {
  return visible ? (
    <Container>
      <CarIcon className="icon-directions_car" />
      {traffic.distance} - {traffic.duration}
    </Container>
  ) : null;
};
