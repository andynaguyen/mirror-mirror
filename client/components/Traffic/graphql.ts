import { graphql } from 'react-apollo';
import gql from 'graphql-tag';
import { TrafficResponse, ApolloWrapperProps } from './types';

const query = gql`
  query {
    traffic {
      distance
      duration
    }
  }
`;

export default graphql<TrafficResponse, ApolloWrapperProps>(query, {
  options: {
    pollInterval: 1000 * 60 * 15, // 15 min
    fetchPolicy: 'network-only',
  },
  props: ({ data: { traffic, error } }) => ({
    traffic: {
      distance: error || !traffic ? 'N/A' : traffic.distance,
      duration: error || !traffic ? 'N/A' : traffic.duration,
    },
  }),
});
