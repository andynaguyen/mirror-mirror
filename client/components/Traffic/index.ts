import Traffic from './Traffic';
import withGraphql from './graphql';

/**
 * A component showing basic traffic conditions updated at every 15 minutes.
 */
const EnhancedTraffic = withGraphql(Traffic);
export default EnhancedTraffic;
