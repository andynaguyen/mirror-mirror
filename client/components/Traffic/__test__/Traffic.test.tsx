import 'jest-styled-components';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import Traffic from '../Traffic';

describe('<Traffic />', () => {
  it('should render the contents of the traffic component', () => {
    const trafficData = { distance: 'N/A', duration: 'N/A' };
    const component = renderer.create(<Traffic traffic={trafficData} visible />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render nothing', () => {
    const trafficData = { distance: 'N/A', duration: 'N/A' };
    const component = renderer.create(<Traffic traffic={trafficData} visible={false} />);
    const tree = component.toJSON();
    expect(tree).toBeNull();
  });
});
