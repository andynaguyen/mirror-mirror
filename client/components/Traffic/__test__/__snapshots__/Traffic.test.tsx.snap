// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Traffic /> should render the contents of the traffic component 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin-top: 8px;
}

.c1 {
  color: white;
  font-size: 22px;
  width: 36px;
}

<div
  className="c0"
>
  <i
    className="icon-directions_car c1"
  />
  N/A
   - 
  N/A
</div>
`;
